File: processor/src/main/java/io/shreyash/rush/blocks/BlockWithParams.java
package io.shreyash.rush.blocks;

import java.util.ArrayList;
import java.util.List;

import javax.lang.model.element.Element;
import javax.lang.model.element.ExecutableElement;

import io.shreyash.rush.util.ConvertToYailType;

public abstract class BlockWithParams extends Block {
  private final ExecutableElement element;

  protected BlockWithParams(Element element) {
    super(element);
    this.element = (ExecutableElement) element;
  }

  /**
   * @return The parameters (or arguments) of this block.
   */
  public List<BlockParam> params() {
    final List<BlockParam> params = new ArrayList<>();

    if (this.element.getParameters() != null) {
      this.element.getParameters().forEach(el -> {
        params.add(new BlockParam(
            el.getSimpleName().toString(), ConvertToYailType.convert(el.asType().toString())
        ));
      });
    }

    return params;
  }
}

class BlockParam {
  // Name of this parameter
  private final String name;

  // YAIL type of this parameter
  private final String type;

  BlockParam(String name, String type) {
    this.name = name;
    this.type = type;
  }

  public String getName() {
    return name;
  }

  public String getType() {
    return type;
  }
}

File: processor/src/main/java/io/shreyash/rush/processor/block/BlockWithParams.kt
package io.shreyash.rush.processor.block

import io.shreyash.rush.processor.util.convert
import javax.lang.model.element.Element

abstract class BlockWithParams protected constructor(element: Element) : Block(element) {
    /**
     * @return The parameters (or arguments) of this block.
     */
    fun params(): List<BlockParam> {
        val params = this.element.parameters
        return params.map {
            BlockParam(it.simpleName.toString(), convert(it.asType().toString()))
        }
    }
}

data class BlockParam(
    // Name of this parameter
    val name: String,
    // YAIL type of this parameter
    val type: String
)
